### PR TITLE
Scan Kernels

### PR DESCRIPTION
**Pull request by MrBurmark**
Add scan kernels.
Add a basic algorithm scan kernel that does an exclusive scan.
Add a basic index list creation kernel that makes an list of indices where a condition is satisfied on a vector. Here it makes a list of indices for negative values so about half of the elements are in the list.

### CODE REVIEW

File: src/algorithm/SCAN-OMP.cpp
Comment (MrBurmark) [inline src/algorithm/SCAN-OMP.cpp:45]: @trws I'm having trouble with the openmp variants where I'm using openmp scan. Clang11 on blueos does not appear to give the right answers for the SCAN kernel here, and it segfaults in the INDEXLIST kernel.

**Comment by MrBurmark** [comment]
The INDEXLIST kernel is implemented with a single loop where this is possible for example in the sequential case, and a setup loop, scan, and final loop where this is not possible like with cuda. Does it make sense to split this into two kernels INDEXSET and INDEXSET_3LOOP so we have both implementations where it is possible to use both?

**Comment by rhornung67** [comment]
> The INDEXLIST kernel is implemented with a single loop where this is possible for example in the sequential case, and a setup loop, scan, and final loop where this is not possible like with cuda. Does it make sense to split this into two kernels INDEXSET and INDEXSET_3LOOP so we have both implementations where it is possible to use both?

I think that makes sense.

File: src/algorithm/SCAN-OMP.cpp
Comment (MrBurmark) [reply src/algorithm/SCAN-OMP.cpp:45]: @trws On a similar not is OpenMP scan supposed to work with target team distribute?

File: src/algorithm/SCAN-Cuda.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-Cuda.cpp:79]: should the startTimer call go up to line 62 for a fair comparison with the RAJA variant?

**Comment by rhornung67** [comment]
@MrBurmark this looks like it's still a work-in-progress. Should I wait until giving a thorough review?

File: src/algorithm/SCAN-Cuda.cpp
Comment (MrBurmark) [reply src/algorithm/SCAN-Cuda.cpp:79]: I agree this is a bit asymmetric, but I didn't want to time the cudaMalloc. The RAJA variant gets memory from a pool that already has memory, but this variant calls cuda Malloc. The RAJA variant queries how much memory it needs each rep, but this variant only queries how much memory it needs once outside the timer. So this variant does do less work each rep than the RAJA variant which is not ideal, but I wasn't sure how to avoid timing the cudaMalloc and get the memory query into the rep loop.

File: src/basic/INDEXLIST-Cuda.cpp
Comment (MrBurmark) [inline src/basic/INDEXLIST-Cuda.cpp:230]: No way to do this with a single kernel/loop with RAJA at the moment.

File: src/basic/INDEXLIST.cpp
Comment (MrBurmark) [inline src/basic/INDEXLIST.cpp:48]: missing hip

File: src/algorithm/SCAN-Cuda.cpp
Comment (rhornung67) [reply src/algorithm/SCAN-Cuda.cpp:79]: OK. I can live with that.

**Comment by MrBurmark** [comment]
@rhornung67 Yes hold off for now, this still needs some work.

File: src/basic/INDEXLIST-Cuda.cpp
Comment (artv3) [reply src/basic/INDEXLIST-Cuda.cpp:230]: It can't be done with RAJA Teams? What would we be missing?

Comment (MrBurmark) [reply src/basic/INDEXLIST-Cuda.cpp:230]: An abstraction for scans that can be done within a single RAJA kernel. We currently do with with 3 loops as can be seen in the INDEXLIST_3LOOP kernel in this PR.

Comment (artv3) [reply src/basic/INDEXLIST-Cuda.cpp:230]: Ah gotcha, a device function of exclusive_scan_inplace, that users can call within their RAJA kernels? that would be useful!

**Review by artv3 — DISMISSED** [review]

File: src/algorithm/SCAN-OMP.cpp
Comment (MrBurmark) [reply src/algorithm/SCAN-OMP.cpp]: I have worked around the openmp scan issues I was seeing by disabling the code with a configure flag. We can revisit this when scan is better supported.

**Comment by rhornung67** [comment]
@MrBurmark should we merge this one next if/when it comes back green?

**Comment by MrBurmark** [comment]
Sure, if people are happy with it.

File: src/algorithm/SCAN-Cuda.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-Cuda.cpp:2]:
```suggestion
// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
```
Please update the copyright year in the new files you added.

File: src/algorithm/SCAN-Hip.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-Hip.cpp:73]: why is this CUB stuff in the HIP file?

File: src/algorithm/SCAN-Hip.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-Hip.cpp:102]: Same question about CUB in HIP file.

File: src/algorithm/SCAN-OMP.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-OMP.cpp:51]: Uggghhh. OpenMP really makes code gross.

File: src/algorithm/SCAN-OMPTarget.cpp
Comment (rhornung67) [inline src/algorithm/SCAN-OMPTarget.cpp:46]: Are these same guards in the SCAN.cpp file where the enabled variants are defined? That will make the output less confusing because it will report that there was no variant run if it's not supported. Otherwise, it will just report zero time and a failed checksum.

Comment (rhornung67) [reply src/algorithm/SCAN-OMPTarget.cpp:46]: Nevermind. I see that you did that. 👍

File: src/algorithm/SCAN-Hip.cpp
Comment (MrBurmark) [reply src/algorithm/SCAN-Hip.cpp:73]: This mirrors what we have in RAJA itself to support building cuda via hip. I'm happy to take it out, of both here and RAJA.

File: src/algorithm/SCAN-OMP.cpp
Comment (MrBurmark) [reply src/algorithm/SCAN-OMP.cpp:51]: yep.

File: src/algorithm/SCAN-OMP.cpp
Comment (rhornung67) [reply src/algorithm/SCAN-OMP.cpp]: @MrBurmark please make an issue so we don't forget to circle back to this.

**Comment by rhornung67** [comment]
> The INDEXLIST kernel is implemented with a single loop where this is possible for example in the sequential case, and a setup loop, scan, and final loop where this is not possible like with cuda. Does it make sense to split this into two kernels INDEXSET and INDEXSET_3LOOP so we have both implementations where it is possible to use both?

I think that's a good idea. You can do it in a different PR if you want.

**Comment by MrBurmark** [comment]
@rhornung67 Looks like corona timed out, can you restart it?

**Comment by MrBurmark** [comment]
> > The INDEXLIST kernel is implemented with a single loop where this is possible for example in the sequential case, and a setup loop, scan, and final loop where this is not possible like with cuda. Does it make sense to split this into two kernels INDEXSET and INDEXSET_3LOOP so we have both implementations where it is possible to use both?
> 
> I think that's a good idea. You can do it in a different PR if you want.

Too late, I added it to this PR.

File: src/algorithm/SCAN-Hip.cpp
Comment (rhornung67) [reply src/algorithm/SCAN-Hip.cpp:73]: Ah, I see. We should see what others think? It's not clear to me what's best.

Comment (MrBurmark) [reply src/algorithm/SCAN-Hip.cpp:73]: Should we add this to the agenda for the next raja meeting?

Comment (rhornung67) [reply src/algorithm/SCAN-Hip.cpp:73]: I'll make a note to add.

**Comment by MrBurmark** [comment]
@rhornung67 No joy, corona timed out again.

**Comment by rhornung67** [comment]
> @rhornung67 No joy, corona timed out again.

I see that. Other projects are impacted as well. I restarted and it passed.

**Pull request merged**
